File: frontend/src/renderer/components/scripts/delete_device.ts

import * as os from 'os';
import * as fs from 'fs';
import * as path from 'path';
import * as dotenv from 'dotenv';
import * as net from 'net';
import * as crypto from 'crypto';
import ConfigParser from 'configparser';

dotenv.config();

const homeDirectory = os.homedir();
const BANBURY_FOLDER = path.join(homeDirectory, '.banbury');
const CONFIG_FILE = path.join(BANBURY_FOLDER, '.banbury_config.ini');

if (!fs.existsSync(BANBURY_FOLDER)) {
    fs.mkdirSync(BANBURY_FOLDER);
}

if (!fs.existsSync(CONFIG_FILE)) {
    const config = new ConfigParser();
    config.set('banbury_cloud', 'credentials_file', 'credentials.json');
    fs.writeFileSync(CONFIG_FILE, config.toString());
}

function connectToRelayServer(): net.Socket {
    const RELAY_HOST = '34.28.13.79';
    const RELAY_PORT = 443;
    const senderSocket = new net.Socket();
    senderSocket.connect(RELAY_PORT, RELAY_HOST);
    return senderSocket;
}

function loadCredentials(): Record<string, string> {
    try {
        const config = new ConfigParser();
        config.read(CONFIG_FILE);
        const credentialsFile = config.get('banbury_cloud', 'credentials_file') || 'default_filename.json';
        const credentialsFilePath = path.join(BANBURY_FOLDER, credentialsFile);
        return JSON.parse(fs.readFileSync(credentialsFilePath, 'utf-8'));
    } catch (error) {
        return {};
    }
}

function saveCredentials(credentials: Record<string, string>): void {
    const config = new ConfigParser();
    config.read(CONFIG_FILE);
    const credentialsFile = config.get('banbury_cloud', 'credentials_file') || 'default_filename.json';
    const credentialsFilePath = path.join(BANBURY_FOLDER, credentialsFile);
    fs.writeFileSync(credentialsFilePath, JSON.stringify(credentials));
}

function delete_file(device_name: string[]): string {
    const senderSocket = connectToRelayServer();
    const endOfHeader = Buffer.from('END_OF_HEADER');
    const credentials = loadCredentials();
    let username = Object.keys(credentials)[0];
    const file_size = ""
    let header: string | null = null;
    let buffer = Buffer.alloc(0);
    let null_arg = ""
    const fileHeader = `DEVICE_DELETE_REQUEST:${device_name}:${null_arg}:${username}:`;
    senderSocket.write(fileHeader);
    senderSocket.write(endOfHeader);

    let jobCompleted = false;
    senderSocket.on('data', (data) => {
        buffer = Buffer.concat([buffer, data]);
        let fileType = 'Unknown';
        if (buffer.includes(endOfHeader) && !header) {
            const endOfHeaderIndex = buffer.indexOf(endOfHeader);
            if (endOfHeaderIndex !== -1) {
                const headerPart = buffer.slice(0, endOfHeaderIndex);
                const content = buffer.slice(endOfHeaderIndex + endOfHeader.length);
                header = headerPart.toString();
                const splitHeader = header.split(':');
                fileType = splitHeader[0];
                buffer = content;
            }
        }
    });

    senderSocket.on('end', () => {
        if (!jobCompleted) {
            console.log('Connection closed before login completion.');
        }
    });

    senderSocket.on('error', (err) => {
        console.error('Error during login:', err);
        senderSocket.end();
    });

    return '';
}


function main(device_name: string[]): void {
      const result = delete_file(device_name);
    }

export default main;

File: frontend/src/main/conservatory/deleteDevice.py
import logging
from flask import Flask,request, jsonify
import sys
from flask_restful import Resource, Api
from flask_cors import CORS, cross_origin
import os
import socket
from dotenv import load_dotenv
import configparser
import json


# Get the home directory of the user
home_directory = os.path.expanduser("~")
# Specify the folder for Banbury Cloud-related files
BANBURY_FOLDER = os.path.join(home_directory, ".banbury")

# Specify the full path to the configuration file
CONFIG_FILE = os.path.join(BANBURY_FOLDER, ".banbury_config.ini")

# Ensure the configuration file exists; create it if it doesn't
if not os.path.exists(CONFIG_FILE):
    # Create a ConfigParser instance with default values
    config = configparser.ConfigParser()
    config["banbury_cloud"] = {
        "credentials_file": "credentials.json"
    }
    
    # Write the configuration to the file
    with open(CONFIG_FILE, "w") as config_file:
        config.write(config_file)


def load_credentials():
    try:
        # Create a ConfigParser instance and read the configuration file
        config = configparser.ConfigParser()
        config.read(CONFIG_FILE)
        credentials_file = config.get("banbury_cloud", "credentials_file")

        # Example: Load credentials from the specified file
        credentials_file_path = os.path.join(BANBURY_FOLDER, credentials_file)
        with open(credentials_file_path, "r") as file:
            return json.load(file)

    except (configparser.Error, FileNotFoundError):
        return {}

def connect_to_relay_server():
    load_dotenv()
    RELAY_HOST = "34.28.13.79" 
    RELAY_PORT = 8002

    sender_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sender_socket.connect((RELAY_HOST, RELAY_PORT))
    return sender_socket



def delete_file(device_name):

    credentials = load_credentials()
    username = next(iter(credentials))    
    sender_socket = connect_to_relay_server()
    print(f"Device Name: {device_name}")
    file_size = ""
    null_arg = ""
    file_header = f"DEVICE_DELETE_REQUEST:{device_name}:{null_arg}:{username}:"
    sender_socket.send(file_header.encode())
    sender_socket.send(b"END_OF_HEADER") # delimiter to notify the server that the header is done

    return

def main():

    if len(sys.argv) > 1:
        device_name = sys.argv[1]
        print(f"Argument received: {device_name}")
    else:
        print("No argument received.")
        device_name = "michael-GF65-Thin-9SD"

    delete_file(device_name)

if __name__ == '__main__':
    main()

